velox/functions/prestosql/aggregates/GeometricMeanAggregate.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cmath>
+#include "velox/exec/Aggregate.h"
+#include "velox/exec/SimpleAggregateAdapter.h"
+#include "velox/functions/prestosql/aggregates/AggregateNames.h"
+
+using namespace facebook::velox::exec;
+
+namespace facebook::velox::aggregate::prestosql {
+
+namespace {
+
+template <typename T>
+class GeometricMeanAggregate {
+ public:
+  using InputType = Row<T>;
+
+  using IntermediateType =
+      Row</*logSum*/ double,
+          /*count*/ int64_t>;
+
+  using OutputType = double;
+
+  static bool toIntermediate(
+      exec::out_type<Row<double, int64_t>>& out,
+      exec::arg_type<T> in) {
+    out.copy_from(std::make_tuple(static_cast<double>(in), 1));
```

Comment:
Can you make sure this code path is covered with tests?

Please, also run AggregationFuzzer and make sure this function is covered well. Use --only geometric_mean --duration_sec 1800

Reply:
It would also be good to test companion functions with AggregationFuzzer since you registered companion functions altogether. i.e., use `--only "geometric_mean,geometric_mean_partial,geometric_mean_merge,geometric_mean_merge_extract" --duration_sec 1800`

Reply:
Two questions here:

1. How to we test `toIntermediate` individually? Is there examples I can take a look at?
2. I tried the AggregationFuzzer, it fails with the following error message:

```
1 extra rows, 1 missing rows
1 of extra rows:
	null | null | null | [-2.6020417535176192,4]

1 of missing rows:
	null | null | null | [-2.602041753517619,4]
```

It seems fails with precision issue? is this expected for floating type functions? Or something wrong with my implementation?

Reply:
> How to we test toIntermediate individually? Is there examples I can take a look at?

AggregationTestBase::testAggregations has logic to test this. See velox/functions/lib/aggregates/tests/AggregationTestBase.cpp

`SCOPED_TRACE("Run partial + final with abandon partial agg");`

However, this logic applies only if input has 2 or more vectors. Hence, to trigger that logic pass 2 or more vectors to testAggregations.

> It seems fails with precision issue?

Fuzzer uses assertEqualResults which has logic to compare with epsilon. Maybe this logic doesn't work for some reason. Would you, please, investigate?

// Compare actualRows with expectedRows and return whether they match. Compare
// actualRows and expectedRows with epsilon if needed and allowed. Otherwise,
// compare their values directly. The underlying assumption is that aggregation
// results can be sorted by unique keys and floating-point values in them are
// computed in different ways and hence require epsilon comparison. For results
// of other operations, floating-point values are likely copied from inputs and
// hence can be compared directly.
bool assertEqualResults(

Reply:
The failure with precision issue is probably due to this: https://github.com/facebookincubator/velox/issues/4481.

To confirm that toIntermediate is tested, you can write a test case as @mbasmanova suggested and set a breakpoint in toIntermediate to make sure the breakpoint gets hit when you run the test.